### DISCUSSION TITLE
Kong open tracing plugin for Elastic APM

### DISCUSSION THREAD

**Discussion by ludovic-pourrat** [Kong/kong#7331] [Ideas and feature requests]
This feature is about developing a Kong plugin to instrument Kong to achieve open tracing with the Elastic APM.

The selected integration is agent-less and leveraging the intake API from Elastic.

See https://www.elastic.co/guide/en/apm/server/current/events-api.html

The implementation design is similar to approach taken by https://github.com/Kong/kong-plugin-zipkin

What is the process for me to contribute in this plugin, as I already have a proof-of-concept working ?

**Comment by ludovic-pourrat** [comment]
![kong-apm](https://user-images.githubusercontent.com/52542520/96574451-c35b4180-12cf-11eb-9e4f-5aeeb2f94016.png)

See open-tracing in action from the current PoC status.

**Comment by guanlan** [comment]
@ludovic-pourrat Thank you for being interested in contributing.

I'd suggest reading the following docs to get you started.
https://github.com/Kong/kong/blob/master/DEVELOPER.md
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md

**Comment by fgsalomon** [comment]
@ludovic-pourrat any updates?  I'm very interested in this plugin, so if there's any repo I could contribute to I'll be glad to help :)

**Reply by ludovic-pourrat** [reply]
Hi,

This was on-hold since few months but now we are restarting this activity.

I can share what I have done so far to include your thoughts ?

**Reply by fgsalomon** [reply]
That would be great!

**Comment by ludovic-pourrat** [comment]
Check this out !

https://github.com/matthyx/kong-elastic-apm

**Reply by zffocussss** [reply]
how to use it?

**Reply by ludovic-pourrat** [reply]
@zffocussss have you found the information needed in the repo ?

**Reply by zffocussss** [reply]
@ludovic-pourrat ,yes,I did.

1. use multiple-stages builds,to copy executable to the docker images of kong (/usr/local/bin/elastic-apm)
2. set environment variables to make it work

- KONG_PLUGINSERVER_ELASTIC_APM_QUERY_CMD:  /usr/local/bin/elastic-apm -dump
- KONG_PLUGINSERVER_NAMES:   elastic-apm
- KONG_PLUGINS: bundled,elastic-apm
